whylog/teacher/__init__.py

```diff
@@ -0,0 +1,78 @@
+class Interval(object):
+    """
+    Represents interval in line.
+    """
+
+    def __init__(self, start_offset, length, abstract_line):
+        self._start_offset = start_offset
+        self._length = length
+        self._abstract_line = abstract_line
+
+    def get_start_offset(self):
+        return self._start_offset
+
+    def get_length(self):
+        return self._length
+
+    def get_location(self):
+        return self._abstract_line
+
+
+class Rule(object):
+    """
+    Represents rule entering by user.
+    """
+
+    def __index__(self, effect):
+        self._effect = effect
+        self._causes = set()
+
+    def add_cause(self, cause):
+        self._causes.add()
+
+    def remove_cause(self, cause):
+        self.causes.remove(cause)
+
+
+class Teacher(object):
+    """
+    Enable teaching new rule. One Teacher per one entering rule.
+    """
+
+    def __init__(self, effect, causes=None):
+        self._rule = Rule(effect)
+        for cause in causes:
```

Comment:
```python
TypeError: iteration over non-sequence
```
when causes is `None` as per default argument.
Anyway, in improved design of `Rule`, `Teacher.causes.update()`